File: java/lib/src/main/java/io/openweb3/walletpay/Currencies.java
package io.openweb3.walletpay;

import io.openweb3.walletpay.exceptions.ApiException;
import io.openweb3.walletpay.internal.api.CurrenciesApi;
import io.openweb3.walletpay.models.Currency;
import io.openweb3.walletpay.models.CursorPageCurrency;

public final class Currencies {
	private final CurrenciesApi api;

	Currencies() {
		api = new CurrenciesApi();
	}

	public CursorPageCurrency list(final CurrencyListOptions options) throws ApiException {
		try {
			return api.v1CurrenciesList(options.getLimit(), options.getCursor(), options.getAppId(), options.getRated());
		} catch (io.openweb3.walletpay.internal.ApiException e) {
			throw Utils.WrapInternalApiException(e);
		}
	}

	// find by code
	public Currency findByCode(final String code) throws ApiException {
		try {
			return api.v1CurrenciesRetrieve(code);
		} catch (io.openweb3.walletpay.internal.ApiException e) {
			throw Utils.WrapInternalApiException(e);
		}
	}
}

File: go/currencies.go
package walletpay

import (
	"context"

	"github.com/openweb3-io/wallet-pay-openapi/go/internal/openapi"
)

type (
	ListResponseCurrencyOut = openapi.CursorPageCurrency
	CurrencyOut             = openapi.Currency
)

type Currencies struct {
	api *openapi.APIClient
}

type CurrencyListOptions struct {
	Cursor *string
	Limit  *int32
	Rated  *bool
}

func (e *Currencies) List(ctx context.Context, options *CurrencyListOptions) (*ListResponseCurrencyOut, error) {
	req := e.api.CurrenciesApi.V1CurrenciesList(ctx)
	if options != nil {
		if options.Cursor != nil {
			req = req.Cursor(*options.Cursor)
		}
		if options.Rated != nil {
			req = req.Rated(*options.Rated)
		}
		if options.Limit != nil {
			req = req.Limit(*options.Limit)
		}
	}
	out, res, err := req.Execute()
	if err != nil {
		return nil, wrapError(err, res)
	}

	ret := ListResponseCurrencyOut(out)
	return &ret, nil
}

func (e *Currencies) FindByCode(ctx context.Context, code string) (*CurrencyOut, error) {
	req := e.api.CurrenciesApi.V1CurrenciesRetrieve(ctx, code)

	out, res, err := req.Execute()
	if err != nil {
		return nil, wrapError(err, res)
	}
	ret := CurrencyOut(out)
	return &ret, nil
}
